devito/dse/aliases.py

```diff
@@ -36,244 +38,246 @@ def collect(exprs):
         a[i+2] - b[i+2] : because at least one operation differs
         a[i+2] + b[i] : because distance along ``i`` differ (+2 and +0)
     """
-    ExprData = namedtuple('ExprData', 'aindices offsets')
-
-    # Discard expressions:
-    # - that surely won't alias to anything
-    # - that are non-scalar
-    candidates = OrderedDict()
+    # Determine the potential aliases
+    candidates = []
     for expr in exprs:
-        if expr.lhs.is_Indexed:
-            continue
-        indexeds = retrieve_indexed(expr.rhs, mode='all')
-        if indexeds and not any(q_indirect(i) for i in indexeds):
-            handle = calculate_offsets(indexeds)
-            if handle:
-                candidates[expr.rhs] = ExprData(*handle)
-
-    aliases = OrderedDict()
-    mapper = OrderedDict()
+        candidate = analyze(expr)
+        if candidate is not None:
+            candidates.append(candidate)
+
+    # Group together the aliasing expressions (ultimately build an Alias for each
+    # group of aliasing expressions)
+    aliases = Aliases()
     unseen = list(candidates)
     while unseen:
```

Comment:
this seems different from the pattern
usually, i expected something like `candidates = list(candidates)` for using it as `while candidates:` in the next line